Match partial results in completion (#1386)

adamwethington7 opened the pull request:
[Issue #1386 ](https://github.com/The-Compiler/qutebrowser/issues/1386)

If there's only one result (I tried it with :bac for :back as Jumblemuddle mentioned), hitting enter will replace the command with the only suggested completion.

adamwethington7 commented:
My bad. I'll get on the build failures.

The-Compiler commented on qutebrowser/commands/runners.py:
```diff
@@ -172,6 +172,20 @@ def parse(self, text, *, aliases=True, fallback=False, keep=False):
                 return self.parse(new_cmd, aliases=False, fallback=fallback,
                                   keep=keep)
         try:
+            
```
Is there a reason this is all inside the `try:`-block? I think it'd make more sense to only catch `KeyError` in the `cmd = cmdutils.cmd_dict[cmdstr]` line, but I might be missing something.

The-Compiler commented on qutebrowser/commands/runners.py:
```diff
@@ -172,6 +172,20 @@ def parse(self, text, *, aliases=True, fallback=False, keep=False):
+                the given command with the match.
+                Ex: If they type "bac" and the only completion is "back", 
+                turn the command into "back".
+            """
```
Please use a normal comment - docstrings are intended as documentations at the start of a function/class

The-Compiler commented on qutebrowser/commands/runners.py:
```diff
@@ -172,6 +172,20 @@ def parse(self, text, *, aliases=True, fallback=False, keep=False):
+
+            matches = []
+            for valid_command in cmdutils.cmd_dict.keys():
+                if valid_command.find(cmdstr) == 0:
```
This can be written as `if cmdstr in valid_command` instead

The-Compiler commented:
Thanks! I added a few line comments.

I'm also wondering if it would be better to add a keyword argument to activate that behaviour to make sure it's only active with the commandline - I think currently you'll get the partial matching even when e.g. spawning commands from a userscript, which I think is undesirable?

adamwethington7 commented:
Hello,

I've fixed the comments and fixed some of the issues that pylint and flake8 pointed out with tox. I moved the code to a different method as pylint was advising that I had too many branches.

As far as the userscripts, I tried a couple of userscripts and had no problems with executing them.

With the if cmdstr in valid_command, I had taken the issue to mean that they wanted it to autocomplete whatever had been typed. If I used if cmdstr in valid_command, it selects string matches in strings that don't begin with what's been typed. If you want, I can change it to if cmdstr in valid_command anyway though, but that's why I had used String.find() instead of using in. 

What can I do about the unexpected coverage changes that codecov reported? I'm fairly new to this.

The-Compiler commented:
> As far as the userscripts, I tried a couple of userscripts and had no problems with executing them.

Yeah, my point is that an userscript now can also do e.g. `echo ba > "$QUTE_FIFO"` and that'll work. I'd prefer to have this behaviour for the interactive commandline only.

> With the if cmdstr in valid_command, I had taken the issue to mean that they wanted it to autocomplete whatever had been typed. If I used if cmdstr in valid_command, it selects string matches in strings that don't begin with what's been typed. If you want, I can change it to if cmdstr in valid_command anyway though, but that's why I had used String.find() instead of using in.

Ah! I think this is equivalent to `if valid_command.startswith(cmdstr):` then? That'd be much clearer :wink:

> What can I do about the unexpected coverage changes that codecov reported? I'm fairly new to this.

You can ignore them. I originally turned off codecov reporting for pull requests as things aren't quite ready for that yet, but it seems like they changed their configuration recently.

The-Compiler commented:
What also would be nice is if you could add a test for this. Something like this in `tests/integration/features/misc.feature` should work:

```gherkin
    Scenario: Partial commandline matching (#1386)
        When I run :message-i "Hello World"
        Then the message "Hello World" should be shown
```

The-Compiler commented:
Any update on this, @adamwethington7 :smile:

The-Compiler commented:
I now merged this with some adjustments:

* 4a7a2e6 Only do partial matching with main CommandParser
* 2f60073 bdd: Allow to run invalid commands via quteproc
* c9d85d3 bdd: Add tests for partial commandline matching
* 5205723 Add unittests for partial command parsing

Thanks!

adamwethington7 commented:
Hello,

I'm sorry for my absence. I had a pretty unexpected situation come up out
of nowhere that needed a lot of my attention. I didn't mean to completely
abandon it like that. Sorry for having to put that burden on you, it wasn't
intentional. Thanks for being kind.

Adam

On Mon, Jun 6, 2016 at 10:24 AM, Florian Bruhin <notifications@github.com>
wrote:

> I now merged this with some adjustments:
>
>    - 4a7a2e6
>    <https://github.com/The-Compiler/qutebrowser/commit/4a7a2e61d3447de5f3463e6d1e1402d8be1278d6>
>    Only do partial matching with main CommandParser
>    - 2f60073
>    <https://github.com/The-Compiler/qutebrowser/commit/2f60073cdf6a6c3f5cca22c4dbf0027ddde4644e>
>    bdd: Allow to run invalid commands via quteproc
>    - c9d85d3
>    <https://github.com/The-Compiler/qutebrowser/commit/c9d85d3a123656a55641d03ce1419e881500e6c4>
>    bdd: Add tests for partial commandline matching
>    - 5205723
>    <https://github.com/The-Compiler/qutebrowser/commit/520572321af2c870fc80eb2b5fe922cc2194a47c>
>    Add unittests for partial command parsing
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/The-Compiler/qutebrowser/pull/1460#issuecomment-223974228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AI6LWfXDuYUArgj0MacvqAQ5H3akVL-gks5qJC2ZgaJpZM4IRWYb>
> .
>

The-Compiler commented:
@adamwethington7 No worries! Thanks for the follow up, I appreciate it. Hope things are looking better (assuming it was the bad kind of unexpected) for you again. :wink:
